// @readalongs/web-component/dist/collection/components/error-message/error-message.js
import { h } from "@stencil/core";
import { MatIcon } from "../../utils/mat-icon";
export class ErrorMessage {
    constructor() {
        /**
         * The on-screen duration of the error message. Zero disables this
         * functionality, the message remains on the screen.
         */
        this.timeout = 0;
        this.timer = 0;
        this.open = true;
    }
    componentDidLoad() {
        if (this.timeout <= 0) {
            return;
        }
        this.timer = setTimeout(() => {
            this.timer = 0;
            this.expired.emit();
            this.open = false;
        }, this.timeout);
    }
    disconnectedCallback() {
        this.open = false;
        if (this.timer) {
            clearTimeout(this.timer);
            this.timer = 0;
        }
    }
    render() {
        if (!this.open) {
            return;
        }
        return (h("p", { "data-test-id": this.data_cy, class: "alert status-error" }, h(MatIcon, null, "error_outline"), " ", this.msg));
    }
    static get is() { return "error-message"; }
    static get styles() { return "error-message {\n      display: block;\n    }"; }
    static get properties() {
        return {
            "msg": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The message to display to the user."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "msg"
            },
            "data_cy": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Cypress data-test-id value."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "data_cy"
            },
            "timeout": {
                "type": "number",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The on-screen duration of the error message. Zero disables this\nfunctionality, the message remains on the screen."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "timeout",
                "defaultValue": "0"
            }
        };
    }
    static get states() {
        return {
            "open": {}
        };
    }
    static get events() {
        return [{
                "method": "expired",
                "name": "expired",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Event get emitted when the timer expires."
                },
                "complexType": {
                    "original": "any",
                    "resolved": "any",
                    "references": {}
                }
            }];
    }
}
